### ISSUE TITLE
Non-overlapping equality check for two Optional values

### ISSUE THREAD

**Issue by sobolevn** [python/mypy#7364]
* Are you reporting a bug, or opening a feature request?

Bug report.

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```ini
[mypy]
strict_optional = True
strict_equality = True
no_implicit_optional = True
```

```python
from typing import Optional

a: Optional[int]
b: Optional[str]

print(a == b)
```

Output:

```
» mypy ex.py
ex.py:6: error: Non-overlapping equality check (left operand type: "Optional[int]", right operand type: "Optional[str]")
```

* What is the actual behavior/output?

`a` can be equal to `b` when they are both `None`. It is allowed by the type definition.

* What is the behavior/output you expect?

That no violation is reported for this line.

* What are the versions of mypy and Python you are using?

- python 3.73
- mypy 0.720

* Do you see the same issue after installing mypy from Git master?

Yes.

**Comment by ilevkivskyi** [comment]
This is actually intentional (see `mypy.checkexpr.ExpressionChecker.dangerous_comparison`), because the underlying code is most likely a bug (or just very fragile code). Technically, everything reported by this flag is not an error, because it will not cause an exception at runtime.

Anyway, we will be happy to reconsider if you will show any realistic use case for this.
